x-element.js

```diff
@@ -1752,9 +1891,20 @@ class TemplateEngine {
     );
   }
 
-  static #insertAllBefore(childNodes, node) {
-    for (const childNode of [...childNodes]) {
-      node.parentNode.insertBefore(childNode, node);
+  static #createCursors(referenceNode) {
+    const startNode = document.createComment('');
+    const node = document.createComment('');
+    referenceNode.parentNode.insertBefore(startNode, referenceNode);
+    referenceNode.parentNode.insertBefore(node, referenceNode);
+    return { startNode, node };
+  }
+
+  static #insertAllBefore(parentNode, referenceNode, nodes) {
+    // Make a copy of the array, else the live NodeList will be mutated as you
+    //  iterate — which would cause us to miss nodes.
+    // Note that passing “null” as the reference node appends nodes to the end.
```

Comment:
I also tried to leave commentary wherever I noticed patterns that might be non-obvious to readers — in this case, certain browser APIs return “live” collections which cannot be iterated over if you plan to do any mutating.